src/basic/scheduler/scheduler_impl.cpp

```diff
@@ -0,0 +1,433 @@
+      uint64_t seq) {
+    Callback cb;
+
+    if (seq != seq_in_process_) {
+      auto it = deferred_.find(seq);
+      if (it != deferred_.end()) {
+        cb.swap(it->second);
+        deferred_.erase(it);
+      } else if (seq_in_process_ != 0) {
+        it = being_processed_.find(seq);
+        if (it != being_processed_.end()) {
+          cb.swap(it->second);
+          being_processed_.erase(it);
+        }
+      }
+    } else {
+      cb.swap(cb_in_process_);
+    }
+
+    return cb;
+  }
+
+  SchedulerImpl::DeferredCallbacks::DeferredCallbacks(SchedulerBackend &backend)
+      : backend_(backend) {}
+
+  void SchedulerImpl::DeferredCallbacks::push(
+      uint64_t seq, Callback cb, bool cancellable,
+      std::weak_ptr<SchedulerBackendFeedback> sch) {
+    if (cancellable) {
+      cancellable_.push(seq, std::move(cb));
+    } else {
+      non_cancellable_.push(seq, std::move(cb));
+    }
+    rescheduleTimer(std::move(sch));
+  }
+
+  void SchedulerImpl::DeferredCallbacks::processNonCancellableItems(
+      gsl::span<Item> items, const std::shared_ptr<Scheduler> &owner) {
+    for (const auto &item : items) {
+      assert(item.cb);
+      item.cb();
+      if (owner.unique()) {
+        break;
+      }
+    }
+  }
+
+  void SchedulerImpl::DeferredCallbacks::processCancellableItems(
+      gsl::span<uint64_t> items, const std::shared_ptr<Scheduler> &owner) {
+    for (const auto &seq : items) {
+      cancellable_.execute(seq);
+
+      if (owner.unique()) {
+        break;
+      }
+    }
+  }
+
+  void SchedulerImpl::DeferredCallbacks::processBothSources(
+      const std::shared_ptr<Scheduler> &owner) {
+    size_t first_cursor = 0;
+    size_t second_cursor = 0;
+
+    for (;;) {
+      // here we guarantee that lower seq number executes earlier,
+      // that's why this function exists
+
+      uint64_t first_seq = current_nc_items_[first_cursor].seq;
+      uint64_t second_seq = current_c_items_[second_cursor];
+
+      if (first_seq < second_seq) {
+        assert(current_nc_items_[first_cursor].cb);
+        current_nc_items_[first_cursor].cb();
+
+        if (owner.unique()) {
+          break;
+        }
+
+        if (++first_cursor >= current_nc_items_.size()) {
+          gsl::span<uint64_t> items(current_c_items_);
+          processCancellableItems(items.subspan(second_cursor), owner);
+          break;
+        }
+      } else {
+        cancellable_.execute(second_seq);
+
+        if (owner.unique()) {
+          break;
+        }
+
+        if (++second_cursor >= current_c_items_.size()) {
+          gsl::span<Item> items(current_nc_items_);
+          processNonCancellableItems(items.subspan(first_cursor), owner);
+          break;
+        }
+      }
+    }
+
+    current_c_items_.clear();
+    current_nc_items_.clear();
+  }
+
+  void SchedulerImpl::DeferredCallbacks::onTimer(
+      std::shared_ptr<Scheduler> owner) {
+    [this, owner = std::move(owner)]() {
```

Comment:
Why it is wrapped into in-place lambda call?

Reply:
lifetime issues